src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs

```diff
@@ -10,19 +13,33 @@ internal partial class LazyItemEvaluator<P, I, M, D>
     {
         class RemoveOperation : LazyItemOperation
         {
-            public RemoveOperation(OperationBuilder builder, LazyItemEvaluator<P, I, M, D> lazyEvaluator)
+            readonly ImmutableList<string> _matchOnMetadata;
+            readonly MatchOnMetadataOptions _matchOnMetadataOptions;
+
+            public RemoveOperation(RemoveOperationBuilder builder, LazyItemEvaluator<P, I, M, D> lazyEvaluator)
                 : base(builder, lazyEvaluator)
             {
+                _matchOnMetadata = builder.MatchOnMetadata.ToImmutable();
+                _matchOnMetadataOptions = builder.MatchOnMetadataOptions;
             }
 
             // todo port the self referencing matching optimization (e.g. <I Remove="@(I)">) from Update to Remove as well. Ideally make one mechanism for both. https://github.com/Microsoft/msbuild/issues/2314
             // todo Perf: do not match against the globs: https://github.com/Microsoft/msbuild/issues/2329
             protected override ImmutableList<I> SelectItems(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
             {
+                ProjectFileErrorUtilities.VerifyThrowInvalidProjectFile(
+                    _matchOnMetadata.IsEmpty || (!_matchOnMetadata.IsEmpty
+                    && _itemSpec.Fragments.Count == 1
```

Comment:
nit:
It might be worth it to take out the conditions between () out and name it.